solver/src/naive_solver/multi_order_solver.rs

```diff
@@ -217,30 +222,32 @@ fn split_into_contexts(
 /// The derivation of this formula is described in https://docs.google.com/document/d/1jS22wxbCqo88fGsqEMZgRQgiAcHlPqxoMw3CJTHst6c/edit
 /// It assumes GP fee (φ) to be 1 and Uniswap fee (Φ) to be 0.997
 ///
-fn compute_uniswap_out(shortage: &TokenContext, excess: &TokenContext) -> U256 {
+fn compute_uniswap_out(shortage: &TokenContext, excess: &TokenContext) -> BigRational {
     let numerator_minuend = 997
-        * (u256_to_bigint(&excess.sell_volume) - u256_to_bigint(&excess.buy_volume))
-        * u256_to_bigint(&shortage.reserve);
+        * (u256_to_big_int(&excess.sell_volume) - u256_to_big_int(&excess.buy_volume))
+        * u256_to_big_int(&shortage.reserve);
     let numerator_subtrahend = 1000
-        * (u256_to_bigint(&shortage.sell_volume) - u256_to_bigint(&shortage.buy_volume))
-        * u256_to_bigint(&excess.reserve);
-    let denominator = (1000 * u256_to_bigint(&excess.reserve))
-        + (997 * (u256_to_bigint(&excess.sell_volume) - u256_to_bigint(&excess.buy_volume)));
-    bigint_to_u256(&((numerator_minuend - numerator_subtrahend) / denominator))
-        .expect("uniswap_out should always be U256 compatible if excess is chosen correctly")
+        * (u256_to_big_int(&shortage.sell_volume) - u256_to_big_int(&shortage.buy_volume))
+        * u256_to_big_int(&excess.reserve);
+    let denominator = (1000 * u256_to_big_int(&excess.reserve))
+        + (997 * (u256_to_big_int(&excess.sell_volume) - u256_to_big_int(&excess.buy_volume)));
+    BigRational::new(numerator_minuend - numerator_subtrahend, denominator)
```

Comment:
Can denominator be zero here?

Reply:
Technically, since 
```
denominator = (1000 * reserve) + (997 * (sell_volume - buy_volume))
```

It looks like there is a whole 2d plane on which this could be zero. One solution, in fact, is when excess reserve, sell and buy volume are all zero.

Reply:
reserve values can not be zero (since it's not possible to extract the entire amount of a Uniswap pool). Therefore denominator can't be zero. I will add a comment in the code...

Reply:
What about the case when `buy_volume` is much larger than `sell_volume` causing the second term here to be negative enough that it negates or nullifies the first term?

Another integer solution to this is 
```
reserve = 997
sell_volume = 1
buy_volume = 1001
```

Reply:
Good point. Maybe it's safe to return a result here. Will fix in a follow up PR